Ziafazal/add pre requisite course via milestones app

ziafazal opened the pull request:
@mattdrayer added pre-requisite feature via milestones app. It would show a drop down for course authors to select from list available courses as a pre-requisite course in "Settings & Details" page. Inside lms it would show the appropriate messaging under side bar of course about page.

cahrens commented:
@mattdrayer The sandbox noted above doesn't seem to work. Please look into it on Monday.

srpearce commented on lms/templates/courseware/course_about.html:
```diff
@@ -319,6 +319,14 @@
+            <i class="icon icon-list-ul"></i>
+            <p class="important-dates-item-title">${_("Requirements")}</p>
+            <span class="important-dates-item-text pre-requisite">${pre_requisite_course}</span>
+            <p class="tip">${_("Access to this course requires the successful completion of ")}${pre_requisite_course}.</p>
```
OK. I'd change that to one of these, then:

You must successfully complete `<course>` before you access this course.

You must successfully complete `<course>` before you access materials for this course.

You must successfully complete `<course>` before you begin this course.

mattdrayer commented on cms/envs/common.py:
```diff
@@ -732,6 +732,16 @@
 
+############## Milestones ##############
+# Turn off milestones app by default
+FEATURES['MILESTONES_APP'] = False
```
The milestones app installation should instead be handled with the OPTIONAL_APPS functionality found above on line  706

cahrens commented:
From @mattdrayer --
Ok, the sandbox is (finally) up and running – you can view the feature at:
Studio:
http://studio.prereqs.m.sandbox.edx.org/settings/details/edX/DemoX/Demo_Course
LMS:
http://prereqs.m.sandbox.edx.org/courses/edX/DemoX/Demo_Course/about
(login: staff@example.com/edx)

cahrens commented on lms/templates/dashboard/_dashboard_course_listing.html:
```diff
@@ -292,6 +292,18 @@ <h4 class="message-title">
+  <% prc_target = reverse('about_course', args=[course_requirements['prc_id'].to_deprecated_string()]) %>
+  <section class="prerequisites">
+    <p class="tip">
+      ${_("You must successfully complete {link_start}{prc_display}{link_end} before you begin this course.").format(
```
This link is not working on the sandbox.

cahrens commented on lms/templates/dashboard/_dashboard_course_listing.html:
```diff
@@ -292,6 +292,18 @@ <h4 class="message-title">
 
 
   </section>
+  % if course_requirements:
```
Is there planned support for multiple pre-reqs?

ziafazal replied:
added backend support for multiple pre-reqs

cahrens commented on lms/templates/dashboard/_dashboard_course_listing.html:
```diff
@@ -292,6 +292,18 @@ <h4 class="message-title">
 
   </section>
+  % if course_requirements:
+  <% prc_target = reverse('about_course', args=[course_requirements['prc_id'].to_deprecated_string()]) %>
```
Don't use to_deprecated_string. Instead unicode(course_requirements['prc_id']). Make this change throughout.

ziafazal replied:
Using unicode instead of  to_deprecated_string now.

cahrens commented on cms/djangoapps/contentstore/tests/test_course_settings.py:
```diff
@@ -243,6 +243,33 @@ def compare_date_fields(self, details, encoded, context, field):
+        self.client.ajax_post(url, course_detail_json)
+        resp = self.client.get_json(url)
+        course_detail_json = json.loads(resp.content)
+        self.assertEqual('', course_detail_json['pre_requisite_course'])
```
I was expecting this to return None, per line 253. Are the empty string and None treated as equivalent states? Would it be possible to just use one?

cahrens commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -744,6 +760,8 @@ def settings_handler(request, course_key_string):
     """
     course_key = CourseKey.from_string(course_key_string)
+    prerequisite_course_enabled = settings.FEATURES.get('ENABLE_PREREQUISITE_COURSES', False)
+    milestones_app_enabled = settings.FEATURES.get('MILESTONES_APP', False)
```
Should a warning be logged if prerequisite_course_enabled is True but milestones_app_enabled is False?

chrisndodge replied:
I'd suggest a hard assert. Better to let operators know that something is misconfigured rather than some obscure error happening.

cahrens commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -768,15 +785,29 @@ def settings_handler(request, course_key_string):
-            })
+            }
+            if prerequisite_course_enabled and milestones_app_enabled:
+                courses, in_process_course_actions = get_courses_accessible_to_user(request)
```
I am having trouble following this code. Can you describe to me what the 2 return values are and how the code below ends up with "course" just being the pre_requesite_courses (on line 795)?

ziafazal replied:
Added backend support for multiple prerequisite courses.

cahrens commented on cms/djangoapps/models/settings/course_details.py:
```diff
@@ -39,6 +39,7 @@ def __init__(self, org, course_id, run):
         self.effort = None  # int hours/week
         self.course_image_name = ""
         self.course_image_asset_path = ""  # URL of the course image
+        self.pre_requisite_course = ""  # pre-requisite course course key
```
Should this be a List to support multiple prereq courses?

ziafazal replied:
Added support for multiple prerequisite courses at backend level.

cahrens commented on cms/envs/common.py:
```diff
@@ -732,6 +732,16 @@
+# Turn off milestones app by default
+FEATURES['MILESTONES_APP'] = False
+if FEATURES['MILESTONES_APP']:
+    INSTALLED_APPS += ('milestones',)
```
Is this configuration necessary? This file imports the LMS common.py.

import lms.envs.common

ziafazal replied:
Opps Did not know that.  Removed from cms.env.common

cahrens commented on lms/templates/dashboard/_dashboard_course_listing.html:
```diff
@@ -292,6 +292,18 @@ <h4 class="message-title">
+      ${_("You must successfully complete {link_start}{prc_display}{link_end} before you begin this course.").format(
+          link_start='<a href="{}">'.format(prc_target),
+          link_end='</a>',
+          prc_display=' '.join([course_requirements['prc_id'].org, course_requirements['prc_id'].course]),
```
Show the course display name. Else what is displayed may not be what the user would actually see. For example, http://studio.prereqs.m.sandbox.edx.org/settings/advanced/ManTestX/ManTest3/2014

ziafazal replied:
Using combination of `display_org_with_default` and `display_number_with_default` as suggested by @marcotuts

cahrens commented on cms/static/js/spec/views/settings/main_spec.js:
```diff
@@ -19,7 +19,8 @@ define([
                 course_image_name : '',
-                course_image_asset_path : ''
+                course_image_asset_path : '',
+                pre_requisite_course : ''
```
Add a test case below for changing this field. Note that when the Settings code was originally written, Jasmine tests were not part of the expectation. They are now, so new fields added to Settings should include unit test coverage.

ziafazal replied:
Added jasmine unit test.

cahrens commented on cms/templates/settings.html:
```diff
@@ -306,6 +306,21 @@ <h2 class="title-2">${_("Requirements")}</h2>
                   <input type="text" class="short time" id="course-effort" placeholder="HH:MM" />
                   <span class="tip tip-inline">${_("Time spent on all course work")}</span>
                 </li>
+                % if settings.FEATURES.get('ENABLE_PREREQUISITE_COURSES') and settings.FEATURES.get('MILESTONES_APP'):
```
What is the workflow for setting this information for courses on edX (where about_page_editable returns False)?

chrisndodge replied:
@cahrens the MVP deliverable is to Open edX which uses the Open edX about page (not Drupal). I believe follow-on work will be done to make the feature work with edx.org (Drupal supported), but that is TBD. @marcotuts am I right?

cahrens commented on cms/templates/settings.html:
```diff
@@ -306,6 +306,21 @@ <h2 class="title-2">${_("Requirements")}</h2>
+                  <form action="#" class="pre-requisite-course-form" method="post">
+                    <label for="pre-requisite-course">${_("Pre-Requisite Course")}</label>
+                    <select class="input" id="pre-requisite-course">
+                        <option value="">None</option>
```
None should be translated (will need to update checks for None also).

cahrens commented on cms/templates/settings.html:
```diff
@@ -306,6 +306,21 @@ <h2 class="title-2">${_("Requirements")}</h2>
+                        <option value="${course_info['course_key']}">${course_info['display_name']}</option>
+                        % endfor
+                    </select>
+                    <span class="tip tip-inline">${_("Minimum level of competency/knowledge prior to taking this course")}</span>
```
This text seems odd to me given that you are specifying a course.

ziafazal replied:
@marcotuts @srpearce can you suggest a better text?

srpearce replied:
Suggestion: Courses that you must complete before you begin this course (this text can be used even if there's just one)

ziafazal replied:
Updated text as suggested.

cahrens commented on common/djangoapps/util/milestones_helpers.py:
```diff
@@ -0,0 +1,136 @@
+    pre_requisite_course_key = None
+    course_milestones = get_course_milestones(course_key=course_key, relationship="requires")
+    if course_milestones:
+        #TODO check if it is valid course key set as namespace
```
Will you take care of this TODO before submitting?

ziafazal replied:
Got rid of these TODOs by validating course keys at contentstore/views/course

cahrens commented on common/djangoapps/util/milestones_helpers.py:
```diff
@@ -0,0 +1,136 @@
+
+def get_prerequisite_course_key(course_key):
+    """
+    Retrieves pre_requisite_course_key for a course from milestones app
```
It looks like this currently returns just the first prerequisite course (if any exist). Update the documentation accordingly?

ziafazal replied:
Updated docs.

cahrens commented on common/djangoapps/util/milestones_helpers.py:
```diff
@@ -0,0 +1,136 @@
+
+def set_prerequisite_course(course_key, prerequisite_course_key_string):
+    """
+    add or update pre-requisite course milestone for a course
```
Make it clear that this method will remove any existing requirements.

A style suggestion-- it is best to make method and class documentation complete sentences.

cahrens commented on common/djangoapps/util/milestones_helpers.py:
```diff
@@ -0,0 +1,136 @@
+
+def add_prerequisite_course(course_key, prerequisite_course_key):
+    """
+    adds pre-requisite course milestone to a course
```
In this case, what happens if prerequisite_course_key_string is "" or None?

ziafazal replied:
I'm relying on the validation(course keys, user) done at milestones app level.

cahrens commented on common/djangoapps/util/milestones_helpers.py:
```diff
@@ -0,0 +1,136 @@
+    milestones = get_courses_milestones(course_ids, relationship="requires")
+    for milestone in milestones:
+        pre_requisite_course_id = milestone['namespace']
+        #TODO check if it is valid course key set as namespace
```
What about this TODO?

ziafazal replied:
This method is now obsolete. Removing it.

cahrens commented on common/djangoapps/util/milestones_helpers.py:
```diff
@@ -0,0 +1,136 @@
+        add_prerequisite_course(course_key, prerequisite_course_key)
+
+
+def get_pre_requisite_courses(course_ids):
```
I don't see that this method is used.

ziafazal replied:
This method is not needed anymore. Removing it.

cahrens commented on common/djangoapps/util/milestones_helpers.py:
```diff
@@ -0,0 +1,136 @@
+                    if key == 'courses' and value:
+                        courses_list = [CourseKey.from_string(course['course_id']) for course in value]
+                        required_courses = required_courses + courses_list
+            # if there are required course then grab first one since a course can have only single pre-requisite course
```
Again, I'm concerned about codifying this current limitation.

ziafazal replied:
Multiple pre-requisite course support is added at backend.

cahrens commented on common/djangoapps/util/milestones_helpers.py:
```diff
@@ -0,0 +1,136 @@
+
+def get_prc_not_completed(user, enrolled_courses):
+    """
+    It would fetch list of prerequisite courses not completed by user
```
Looks like it actually returns a dict. And the code itself says it will only return a single pre-requisite course.

ziafazal replied:
Updated doc string.

cahrens commented on common/djangoapps/util/milestones_helpers.py:
```diff
@@ -0,0 +1,136 @@
+    pre_requisite_courses = {}
+    if settings.FEATURES.get('MILESTONES_APP') and settings.FEATURES.get('ENABLE_PREREQUISITE_COURSES'):
+        for course_id in enrolled_courses:
+            ful_fillment_paths = get_course_milestones_fulfillment_paths(course_id, {'id': user.id})
```
typo: ful_fillment_paths.

cahrens commented on common/djangoapps/util/milestones_helpers.py:
```diff
@@ -0,0 +1,136 @@
+            and settings.FEATURES.get('ENABLE_PREREQUISITE_COURSES', False) \
+            and course_descriptor.pre_requisite_course:
+        pre_requisite_course_key = CourseKey.from_string(course_descriptor.pre_requisite_course)
+        pre_requisite_course_display = ' '.join([pre_requisite_course_key.org, pre_requisite_course_key.course])
```
I think the display name should be included.

marcotuts replied:
@explorerleslie  any thoughts on whether just using the course ID here is the level of context we'd like to provide for course prerequisite information? 

(See http://prereqs.m.sandbox.edx.org/courses/edX/DemoX/Demo_Course/about for the Requirements section in the right sidebar.) 

My original thought here was that we would want to avoid confusion around similarly named courses, or cut off longer course names. It is also often comment for course handbooks to list prerequisites using course IDs (as course titles might change more frequently for marketing reasons but IDs may stay more stable.) 

We could also leave the ID in the table but add the course title to the sentence included in the requirements section as well. 

Thanks @cahrens for noting this!

explorerleslie replied:
@marcotuts @cahrens I leave it to the solutions team to make the final decision, but I agree with the direction implemented, with the course ID. It's probably easier to search the course catalog via the course ID than the name. However, @marcotuts keep in mind that when Split modulestore is rolled out very soon that course IDs are no longer required to be unique. You may need to add the run in there somehow.

marcotuts replied:
@ziafazal, @mattdrayer - while reviewing this @explorerleslie noted that two advanced settings exists that could trump the rendered string for students for the course ID.  See the following advanced settings in Studio : 
![image](https://cloud.githubusercontent.com/assets/2023680/5620502/83cc9f9a-94fa-11e4-9086-1b4603b44a41.png)

cahrens commented on common/djangoapps/util/milestones_helpers.py:
```diff
@@ -0,0 +1,136 @@
+
+from collections import defaultdict
+from opaque_keys.edx.keys import CourseKey
+from milestones.api import (
```
All of these methods should have unit tests. https://jenkins.testeng.edx.org/job/edx-all-tests-manual-pr/1313/SHARD=1,TEST_SUITE=unit/artifact/reports/diff_coverage_combined.html

ziafazal replied:
Added more unit tests to improve coverage.

cahrens commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -184,6 +184,11 @@ class CourseFields(object):
         help=_("Enter the date you want to advertise as the course start date, if this date is different from the set start date. To advertise the set start date, enter null."),
         scope=Scope.settings
     )
+    pre_requisite_course = String(
```
Sorry to keep saying the same thing over and over... but.... changing this from a String to a List later will be difficult. Why not make this a List now? FYI, @mattdrayer

ziafazal replied:
Multiple pre-requisite course support is added at backend.

cahrens commented on common/test/acceptance/tests/helpers.py:
```diff
@@ -167,6 +168,31 @@ def enable_css_animations(page):
+    return CourseLocator(org, number, run, deprecated=(default_store == 'draft'))
+
+
+def select_ddl_by_value(browser_query, value):
```
What does "DDL" stand for? In general we try to avoid abbreviations.

ziafazal replied:
Changed method names as suggested by @cahrens

cahrens commented on common/test/acceptance/tests/lms/test_lms.py:
```diff
@@ -601,6 +608,85 @@ def test_tooltip(self):
+
+    def test_dashaboad_message(self):
+        """
+        Verify that pre-requisite messages are displayed.
```
A best practice for bok choy tests is to include a BDD description of the test in the comments. For example, 
https://github.com/edx/edx-platform/blob/master/common/test/acceptance/tests/lms/test_lms_instructor_dashboard.py#L40

cahrens commented:
@mattdrayer I am done with my first pass through this PR. Let me know when it is ready for re-review (with updated sandbox).

mattdrayer commented:
Great, thanks for taking the time!  We'll address these items and let you
know when the new sandbox is up.

On Mon, Jan 5, 2015 at 1:36 PM, Christina Roberts <notifications@github.com>
wrote:

> @mattdrayer <https://github.com/mattdrayer> I am done with my first pass
> through this PR. Let me know when it is ready for re-review (with updated
> sandbox).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/6126#issuecomment-68753256>.
>



-- 
Matthew Drayer
Solutions Architect
edX | MIT
mattdrayer@edx.org

mattdrayer commented:
@ziafazal -- can you please include additional unit tests during your next cycle -- there are some modules which require further coverage -- see https://jenkins.testeng.edx.org/job/edx-all-tests-manual-pr/1303/SHARD=1,TEST_SUITE=unit/artifact/reports/diff_coverage_combined.html

chrisndodge commented on cms/static/js/models/settings/course_details.js:
```diff
@@ -15,7 +15,8 @@ var CourseDetails = Backbone.Model.extend({
         course_image_name: '', // the filename
-        course_image_asset_path: '' // the full URL (/c4x/org/course/num/asset/filename)
+        course_image_asset_path: '', // the full URL (/c4x/org/course/num/asset/filename)
+        pre_requisite_course: ''
```
As per @cahrens earlier suggestion to get things ready for multiple pre-requisites, perhaps this should be an array and - for now - assert that len(array) is 1 on the handlers

ziafazal replied:
Added multiple pre-requisite course support at backend.

srpearce commented on lms/templates/courseware/course_about.html:
```diff
@@ -319,6 +319,14 @@
+          % if pre_requisite_course:
+          <li class="prerequisite-course important-dates-item">
+            <i class="icon icon-list-ul"></i>
+            <p class="important-dates-item-title">${_("Requirements")}</p>
```
Just thought of this: Are we listing course requirements (general skills and knowledge) and course prerequisites (specific courses the student has to pass) separately? If so, this heading should be "Prerequisites" or "Course Prerequisites".

ziafazal replied:
@srpearce "Prerequisites" area in the body of the page is coming from "Overview" section of "Sechdule & Details" page on studio. so I can not change that text However, I can change "Requirements" text on the sidebar to "Prerequisites".

ziafazal replied:
Oops there is already a "Prerequisites" area in sidebar.

srpearce replied:
@ziafazal I'm only seeing a **Requirements** section, and it contains what we're calling a prerequisite - ?
![screenshot 228](https://cloud.githubusercontent.com/assets/3493253/5647162/646991be-9651-11e4-9ac0-2f5a7834e241.png)

srpearce replied:
@marcotuts ^^

Also, as Zia pointed out, the "Prerequisites" heading in the body of the About page comes from the default HTML in the **Course Overview** box on the **Schedule & Details** page. Can we change that default HTML so that it says "Requirements/Add information about the skills and knowledge students need to take this course" instead of "Prerequisites/Add information about course prerequisites here"?

![screenshot-229](https://cloud.githubusercontent.com/assets/3493253/5648942/65161d80-9660-11e4-968c-605724178a0e.png)

chrisndodge commented on common/djangoapps/util/milestones_helpers.py:
```diff
@@ -0,0 +1,136 @@
+    milestone = add_milestone({
+        'name': 'Course {} requires {}'.format(unicode(course_key), unicode(prerequisite_course_key)),
+        'namespace': unicode(prerequisite_course_key),
+        'description': '',
```
Since this is a system defined milestone, should we have a boiler-plate description here (and localized)?

chrisndodge commented on common/djangoapps/util/milestones_helpers.py:
```diff
@@ -0,0 +1,136 @@
+    add_course_milestone(course_key, 'requires', milestone)
+
+    # add fulfillment course milestone
+    add_course_milestone(prerequisite_course_key, 'fulfills', milestone)
```
As I suggested to @mattdrayer I'd like to suggest changing these hard-coded relationship type strings to be explicitly defined (in edx-milestones) as constants.

cahrens commented on common/test/acceptance/tests/helpers.py:
```diff
@@ -167,6 +168,31 @@ def enable_css_animations(page):
+
+def select_ddl_by_value(browser_query, value):
+    """
+    Selects a html select element by matching value attribute
```
You inspired me to create some Select helper methods in my own PR. We should try to keep method names consistent. So "select_option_by_value" and "is_option_value_selected"?

https://github.com/edx/edx-platform/pull/6308/files#diff-d2fd85d338b26adbce4621586d16300dR173

ziafazal replied:
Will change the names as you suggested.

srpearce commented on cms/templates/settings.html:
```diff
@@ -306,6 +306,21 @@ <h2 class="title-2">${_("Requirements")}</h2>
+                % if settings.FEATURES.get('ENABLE_PREREQUISITE_COURSES') and settings.FEATURES.get('MILESTONES_APP'):
+                <li class="field field-select" id="field-pre-requisite-course">
+                  <form action="#" class="pre-requisite-course-form" method="post">
+                    <label for="pre-requisite-course">${_("Pre-Requisite Course")}</label>
```
Change to "Prerequisite Course" (no hyphen)

chrisndodge commented:
@ziafazal @mattdrayer sorry one more comment. I think I asked this before, but as we're getting closer to the deadline I'll ask again: how are we setting the fulfillment of the prerequisite milestone? I believe the design is to tie into course end manual processes regarding certificate generation. Do we have that work completed or is that still TBD?

srpearce commented:
@marcotuts Recapping the conversation we just had:

To keep our terminology consistent, we should use "prerequisites" to refer to specific courses that a student has to take before signing up for a course, and "requirements" to refer to general skills and knowledge a student needs*. To reflect that terminology, we need to make two changes:
- In the Course About page sidebar, change the word "Requirements" to "Prerequisites".
![screenshot 228](https://cloud.githubusercontent.com/assets/3493253/5650046/c78aabfe-9668-11e4-9b26-133f041e7767.png)

- In the body of the Course About page, change "Prerequisites/Add information about course prerequisites here" to "Requirements/Add information about the skills and knowledge students need to take this course". To do this, the default template in Studio for About pages needs a minor modification.
![screenshot-229](https://cloud.githubusercontent.com/assets/3493253/5650043/c1ee0e98-9668-11e4-8d63-aea7a3a1534c.png)

*I know that we briefly discussed using these terms in the opposite way ("prereqs"=general skills and "requirements"=specific courses), but I think we decided to use "prereqs" to mean specific courses since that's how we've been using this term all along in conjunction with this project. Correct?

ziafazal commented:
@mattdrayer @chrisndodge @cahrens @marcotuts @srpearce Addressed most of issues raised in feedback. Changes label on course about page for Prerequisites and updated the default Requirements section in course overview HTML.

cahrens commented:
There are a lot of bok choy test failures. I don't know if something went awry in the Jenkins test run, or if these are actual issues. Can someone look into them?

@mattdrayer let me know when this PR is ready for re-review. Please make sure there is an updated sandbox.

chrisndodge commented on cms/templates/settings.html:
```diff
@@ -308,6 +308,21 @@ <h2 class="title-2">${_("Requirements")}</h2>
+                    <label for="pre-requisite-course">${_("Prerequisite Course")}</label>
+                    <select class="input" id="pre-requisite-course">
+                        <option value="">${_("None")}</option>
+                        % for course_info in sorted(pre_requisite_courses, key=lambda s: s['display_name'].lower() if s['display_name'] is not None else ''):
```
I just noticed that we're showing the display name in the drop down list. I'm wondering that if there are two runs of the same course, it's possible that authors will keep the same course display name between runs and this may cause a situation where the DDL is ambiguous.

It's probably not important now, but maybe something to do as a follow-up release

@marcotuts?

chrisndodge commented on lms/djangoapps/certificates/models.py:
```diff
@@ -118,6 +122,17 @@ def certificate_for_student(cls, student, course_id):
+    if settings.FEATURES.get('ENABLE_PREREQUISITE_COURSES') and instance.status == CertificateStatuses.generating:
+        fulfill_course_milestone(instance.course_id, instance.user)
+
+
```
@mattdrayer seems like the fulfillment PR got merged into this branch, given that I see this here...

chrisndodge commented on lms/templates/courseware/course_about.html:
```diff
@@ -319,8 +319,16 @@
+          <li class="prerequisite-course important-dates-item">
+            <i class="icon fa fa-list-ul"></i>
+            <p class="important-dates-item-title">${_("Prerequisites")}</p>
+            <span class="important-dates-item-text pre-requisite">${pre_requisite_courses[0]}</span>
```
nitpick: for now we might want a comment to explain why we're just dereferencing the first element

chrisndodge commented on lms/templates/dashboard/_dashboard_course_listing.html:
```diff
@@ -309,6 +309,18 @@ <h4 class="message-title">
+      ${_("You must successfully complete {link_start}{prc_display}{link_end} before you begin this course.").format(
+          link_start='<a href="{}">'.format(prc_target),
+          link_end='</a>',
+          prc_display=course_requirements['courses'][0]['display'],
```
same nitpick comment re: explaining to code readers why we're just dereferencing the first element (for now)

chrisndodge commented:
Went through the code. I don't have a lot of additional comments exception for one observation about the display names in the DDL.

It does seem like the features added in https://github.com/edx/edx-platform/pull/6480 (which was a branch off of this) is already included. @ziafazal Can we confirm that everything is in this branch that we expect?

Has the sandbox been refreshed so we can try it out?

ziafazal commented:
@chrisndodge @mattdrayer I have merged fulfilment branch into this one So now we have everything related to PRC in this branch.

cahrens commented:
A request for future reviews-- once a review is in process, please don't squash commits that have already been reviewed with ones that have not been seen by all the reviewers. It would have been easier for me to just review the changes from the last time that I looked at the PR, but because everything has been squashed into a single commit, I need to re-review the whole PR.

After code review is complete, then the author can squash the commits before merging the PR.

Thank you!

cahrens commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -469,6 +440,52 @@ def course_index(request, course_key):
+
+def _remove_in_process_courses(courses, in_process_course_actions):
+    """
+    remove any courses in courses that are also in the in_process_course_actions list and formats them for view
```
Nit: I know you just refactored this code, but it would be helpful for you to add to the documentation what an "in_process_course" is. Both times I read this PR I got confused because I thought those are the courses the user is in the process of taking. But looking at _accessible_courses_list_from_groups, I see that "in-process" actually refers to courses that are in the process of being generated for re-run.

cahrens commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -768,15 +785,31 @@ def settings_handler(request, course_key_string):
+                if prerequisite_course_enabled:
+                    prerequisite_course_keys = request.json.get('pre_requisite_courses', [])
+                    if not all(is_valid_course_key(course_key) for course_key in prerequisite_course_keys):
+                        return JsonResponseBadRequest({_("error"): _("Invalid prerequisite course key")})
```
I don't think "error" should be translated. That is a key for the user interface to retrieve the error message.

cahrens commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -768,15 +785,31 @@ def settings_handler(request, course_key_string):
+                courses = [course for course in courses if course.id != course_key]
+                if courses:
+                    courses = _remove_in_process_courses(courses, in_process_course_actions)
+                settings_context.update({'pre_requisite_courses': courses})
```
Nit: these aren't the pre_requisite courses for this course. They are the list of all possible pre_requisite courses. Naming the variable that way would make the code easier to understand.

ziafazal replied:
Renamed `pre_requisite_courses` to `possible_pre_requisite_courses`

cahrens commented on common/djangoapps/util/milestones_helpers.py:
```diff
@@ -0,0 +1,151 @@
+        'description': _('System defined milestone'),
+    })
+    # add requirement course milestone
+    add_course_milestone(course_key, 'requires', milestone)
```
Will these hard-coded strings be changed to constants as @chrisndodge suggested in a future PR?

ziafazal replied:
Yeah. We will replace once milestones app has constants.

mattdrayer replied:
Constant values are currently available in the release_review branch via api.get_milestone_relationship_types()

cahrens commented on common/djangoapps/util/milestones_helpers.py:
```diff
@@ -0,0 +1,151 @@
+    if settings.FEATURES.get('ENABLE_PREREQUISITE_COURSES', False) and course_descriptor.pre_requisite_courses:
+        for course_id in course_descriptor.pre_requisite_courses:
+            course_key = CourseKey.from_string(course_id)
+            required_course_descriptor = modulestore().get_course(course_key)
```
Create a helper method that returns the display name given a course_key. Same code appears above, lines 91-97.

ziafazal replied:
Created helper method.

cahrens commented on common/djangoapps/util/milestones_helpers.py:
```diff
@@ -0,0 +1,151 @@
+
+def fulfill_course_milestone(course_key, user):
+    """
+    It would save course milestone collected by user.
```
I don't understand what this method is doing. Is that what you call when a user completes a course? If so, something like this would be clearer:

Marks the course specified by the given course_key as complete for the given user. If any other courses require this course as a prerequisite, their milestones will be appropriately updated.

ziafazal replied:
Changed doctring as you suggested.

cahrens commented on common/djangoapps/util/milestones_helpers.py:
```diff
@@ -0,0 +1,151 @@
+    validates course key. returns True if valid else False.
+    """
+    try:
+        course_key = CourseKey.from_string(key)
```
@dmitchell is it necessary to go through all of this to validate a course key (is there an existing validation method)? I thought CourseKey.from_string would handle route to SlashSeparatedCourseKey.from_deprecated_string if necessary.

cahrens commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -184,6 +184,11 @@ class CourseFields(object):
         help=_("Enter the date you want to advertise as the course start date, if this date is different from the set start date. To advertise the set start date, enter null."),
         scope=Scope.settings
     )
+    pre_requisite_courses = List(
```
This is showing up in Studio Advanced Settings. 
![image](https://cloud.githubusercontent.com/assets/484484/5682786/97aa908c-97f0-11e4-9e50-bc2c808d4ecb.png)

You need to add this setting name to the list filtered out of Advanced Settings in course_metadata.py (FILTERED_LIST).

ziafazal replied:
added `pre_requisite_courses` to FILTERED_LIST.

cahrens commented on lms/djangoapps/courseware/access.py:
```diff
@@ -323,6 +324,14 @@ def can_load():
+        if settings.FEATURES['ENABLE_PREREQUISITE_COURSES'] \
+                and not _has_staff_access_to_descriptor(user, descriptor, course_key) \
+                and hasattr(descriptor, 'pre_requisite_courses') and descriptor.pre_requisite_courses \
+                and not user.is_anonymous() \
```
Why do we want to allow anonymous access to something that has prereqs?

ziafazal replied:
we are not allowing anonymous access we are don't want to pull fulfillments of anonymous user since we don't have `id` attribute of user.

cahrens commented on common/djangoapps/util/milestones_helpers.py:
```diff
@@ -0,0 +1,151 @@
+    """
+    pre_requisite_courses = {}
+    if settings.FEATURES.get('ENABLE_PREREQUISITE_COURSES'):
+        for course, __ in enrolled_courses:
```
What is this second field that is part of enrolled_courses?

ziafazal replied:
`enrolled_courses` is a list returned by `get_course_enrollment_pairs` method in student/views.py. I'm now filtering this list to get only course keys.

cahrens commented on lms/djangoapps/courseware/access.py:
```diff
@@ -323,6 +324,14 @@ def can_load():
+                and not _has_staff_access_to_descriptor(user, descriptor, course_key) \
+                and hasattr(descriptor, 'pre_requisite_courses') and descriptor.pre_requisite_courses \
+                and not user.is_anonymous() \
+                and get_pre_requisite_courses_not_completed(user, [(descriptor, None)]):
```
get_pre_requisite_courses_not_completed checks if descriptor.pre_requisist_courses is "truthy", so I don't think you also nee the check on line 330.

Why is the "None" passed here?

ziafazal replied:
passing a list of course_keys user is enrolled in to `get_pre_requisite_courses_not_completed` instead of list of tuples.

cahrens commented:
@mattdrayer and @cdodge If I set a prerequisite course for "Course A", then I can no longer see "Course A's" about page. That seems wrong.

The access checking is currently in _has_access_descriptor. This means you can't see any portion of the course if a prerequisite is not complete. Historically we have allowed access to certain portions of the course (for instance about page and course info page) even if the user is not enrolled.

chrisndodge commented:
@cahrens that's probably a very good catch.

My gut says is that it could be due to the permissions change in access.py

cahrens commented:
I am done with this pass through the PR. Let me know when it is ready for re-review, with updated sandbox.

chrisndodge commented:
I'd also be curious to see if the unfulfilled course is removed from the course catalog as well as I think it also goes through the can_load() authZ check.

ziafazal commented:
@mattdrayer @chrisndodge @cahrens  created a new permission 'load_with_prerequisites' to check if user can load a course which has prerequisites.

cahrens commented:
@cdodge have you taken a look at the proposed permissions change? @mattdrayer is there an updated sandbox?

chrisndodge commented:
@cahrens I'm going to come up with a suggested approach for @ziafazal to take. As I mentioned before, it's very similar to what White Label experienced to support 'private courses'. I do believe there's a design flaw in this course visibility rule in Open edX, but addressing that might be best served outside of this PR.

cahrens commented:
OK, I will wait to re-review until the dust settles (since I imagine some portion of commit 103da52 will need to be removed).

chrisndodge commented:
@cahrens I did a small tweek on @ziafazal behalf (renamed the new permission and added a few comments as to what is happening)

cahrens commented:
@chrisndodge test failures.

chrisndodge commented:
@cahrens hrm... not sure what happened there. I ran the full suite of LMS tests locally, but apparently it didn't run them all (is there a side effect to using --fasttest?!?). It seemed like it did complete very quickly... :-)

Give this a chance to run through a Jenkins pass....

cahrens commented:
Merge conflicts now. Content Libraries merged to master, so I can imagine a lot of rebasing going on.

cahrens commented:
Did someone verify that the course catalog access is working properly?

chrisndodge commented:
@cahrens @ziafazal @bradenmacdonald 

There's two merge conflicts, one of which is trivial.

However, there is a merge conflict (due to content libraries landing into master) on cms/djangoapps/contentstore/course.py which I don't feel confident to resolve myself since I'm not one of the authors.

@ziafazal can you try to resolve and tag @bradenmacdonald if you have any questions about the conflict resolution? Thanks

ziafazal commented:
@chrisndodge @cahrens @bradenmacdonald merged with master.

andy-armstrong commented:
@ziafazal What do you mean 'merged with master'? It doesn't look like this PR has been merged on GitHub. Also there are still quite a few PyLint and PEP8 violations that need to be fixed as they are breaking the build.

andy-armstrong commented:
@ziafazal I pulled the latest changes from master and I'm not seeing any of these commits there yet.

andy-armstrong commented:
@ziafazal In case you're not familiar with the Jenkins output, this console log shows the PEP8/PyLint failures:

https://jenkins.testeng.edx.org/job/edx-all-tests-auto-pr/21726/SHARD=1,TEST_SUITE=quality/console

ziafazal commented:
@andy-armstrong sorry I should have said "Rebased with mater" and resolved conflicts. Let me fix the quality errors.

cahrens commented:
There is a bok choy failure in a test related to this PR.

mattdrayer commented:
Yes, I'm trying to get a hand from the testeng folks on this error -- it
has something to do with the way the test database is set up during the bok
choy run -- the milestones test requires some seed data to be present in
order to succeed.

On Wed, Jan 14, 2015 at 10:12 AM, Christina Roberts <
notifications@github.com> wrote:

> There is a bok choy failure in a test related to this PR.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/6126#issuecomment-69930020>.
>



-- 
Matthew Drayer
Solutions Architect
edX | MIT
mattdrayer@edx.org

cahrens commented:
:+1: after the following:

1) Test for course catalog working
2) Bok choy test passing
3) Squash of commits into 1 or relatively small number of logical commits

mattdrayer commented:
All validation suites passed, merging feature
